File: BookLite/src/main/java/org/cubexmc/booklite/storage/BookRepository.java
package org.cubexmc.booklite.storage;

import java.io.File;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Types;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import org.cubexmc.booklite.BookLitePlugin;
import org.cubexmc.booklite.config.ConfigManager;
import org.cubexmc.booklite.model.BookRecord;

public class BookRepository {

    private static final Gson GSON = new Gson();

    private final BookLitePlugin plugin;
    private final ConfigManager config;
    private Connection connection;

    public BookRepository(BookLitePlugin plugin, ConfigManager config) {
        this.plugin = plugin;
        this.config = config;
    }

    public synchronized void init() {
        try {
            Class.forName("org.sqlite.JDBC");
            if (!plugin.getDataFolder().exists() && !plugin.getDataFolder().mkdirs()) {
                plugin.getLogger().warning("Cannot create BookLite data folder.");
            }
            File db = new File(plugin.getDataFolder(), config.getSqliteFile());
            connection = DriverManager.getConnection("jdbc:sqlite:" + db.getAbsolutePath());
            try (Statement st = connection.createStatement()) {
                st.execute("PRAGMA busy_timeout = 5000");
                if (config.isWal()) {
                    st.execute("PRAGMA journal_mode = WAL");
                    st.execute("PRAGMA synchronous = NORMAL");
                }
                st.execute("PRAGMA foreign_keys = ON");
                st.execute("""
                        CREATE TABLE IF NOT EXISTS books (
                          id TEXT PRIMARY KEY,
                          content_hash TEXT NOT NULL,
                          title_raw TEXT NOT NULL,
                          author TEXT NOT NULL,
                          pages_json TEXT NOT NULL,
                          total_pages INTEGER NOT NULL,
                          created_at INTEGER NOT NULL,
                          updated_at INTEGER NOT NULL,
                          last_accessed_at INTEGER NULL
                        )
                        """);
                ensureColumn(st, "last_accessed_at",
                        "ALTER TABLE books ADD COLUMN last_accessed_at INTEGER NULL");
                dropDeletedAtColumn(st);
                st.execute("CREATE INDEX IF NOT EXISTS idx_books_hash ON books(content_hash)");
                st.execute("CREATE INDEX IF NOT EXISTS idx_books_created ON books(created_at)");
                st.execute("CREATE INDEX IF NOT EXISTS idx_books_last_accessed ON books(last_accessed_at)");
                st.execute("PRAGMA user_version = 3");
            }
        } catch (Exception ex) {
            plugin.getLogger().log(Level.SEVERE, "Failed to initialize BookLite SQLite storage.", ex);
            throw new IllegalStateException(ex);
        }
    }

    public synchronized void close() {
        if (connection == null) return;
        try {
            connection.close();
        } catch (SQLException ex) {
            plugin.getLogger().log(Level.WARNING, "Failed to close BookLite database.", ex);
        }
    }

    public synchronized BookRecord saveOrGet(BookRecord draft) throws SQLException {
        BookRecord existing = findByHash(draft.contentHash());
        if (existing != null) return existing;

        try (PreparedStatement ps = connection.prepareStatement("""
                INSERT INTO books
                  (id, content_hash, title_raw, author, pages_json, total_pages, created_at, updated_at, last_accessed_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                """)) {
            ps.setString(1, draft.id());
            ps.setString(2, draft.contentHash());
            ps.setString(3, draft.title());
            ps.setString(4, draft.author());
            ps.setString(5, GSON.toJson(draft.pages()));
            ps.setInt(6, draft.totalPages());
            ps.setLong(7, draft.createdAt());
            ps.setLong(8, draft.updatedAt());
            if (draft.lastAccessedAt() == null) {
                ps.setNull(9, Types.BIGINT);
            } else {
                ps.setLong(9, draft.lastAccessedAt());
            }
            ps.executeUpdate();
        }
        return draft;
    }

    public synchronized BookRecord find(String id) throws SQLException {
        if (id == null || id.isBlank()) return null;
        try (PreparedStatement ps = connection.prepareStatement("SELECT * FROM books WHERE id = ?")) {
            ps.setString(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next() ? read(rs) : null;
            }
        }
    }

    public synchronized List<BookRecord> findByPrefix(String prefix, int limit) throws SQLException {
        List<BookRecord> out = new ArrayList<>();
        if (prefix == null || prefix.isBlank()) return out;
        try (PreparedStatement ps = connection.prepareStatement(
                "SELECT * FROM books WHERE id LIKE ? ORDER BY created_at DESC LIMIT ?")) {
            ps.setString(1, prefix + "%");
            ps.setInt(2, Math.max(1, limit));
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    out.add(read(rs));
                }
            }
        }
        return out;
    }

    public synchronized List<String> completeIdsByPrefix(String prefix, int limit) throws SQLException {
        List<String> out = new ArrayList<>();
        try (PreparedStatement ps = connection.prepareStatement(
                "SELECT id FROM books WHERE id LIKE ? ORDER BY created_at DESC LIMIT ?")) {
            ps.setString(1, (prefix == null ? "" : prefix.toLowerCase()) + "%");
            ps.setInt(2, Math.max(1, limit));
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    out.add(rs.getString("id"));
                }
            }
        }
        return out;
    }

    public synchronized List<BookRecord> list(int offset, int limit) throws SQLException {
        List<BookRecord> out = new ArrayList<>();
        try (PreparedStatement ps = connection.prepareStatement(
                "SELECT * FROM books ORDER BY created_at DESC LIMIT ? OFFSET ?")) {
            ps.setInt(1, Math.max(1, limit));
            ps.setInt(2, Math.max(0, offset));
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    out.add(read(rs));
                }
            }
        }
        return out;
    }

    public synchronized boolean delete(String id) throws SQLException {
        try (PreparedStatement ps = connection.prepareStatement(
                "DELETE FROM books WHERE id = ?")) {
            ps.setString(1, id);
            return ps.executeUpdate() > 0;
        }
    }

    public synchronized boolean markAccessed(String id, long now) throws SQLException {
        try (PreparedStatement ps = connection.prepareStatement(
                "UPDATE books SET last_accessed_at = ?, updated_at = ? WHERE id = ?")) {
            ps.setLong(1, now);
            ps.setLong(2, now);
            ps.setString(3, id);
            return ps.executeUpdate() > 0;
        }
    }

    /**
     * Permanently deletes records whose last player access (or, if never accessed,
     * creation time) is older than {@code staleBefore}. Admin operations like
     * /booklite read do not touch last_accessed_at, so they don't keep records alive.
     */
    public synchronized int purgeStale(long staleBefore) throws SQLException {
        try (PreparedStatement ps = connection.prepareStatement(
                "DELETE FROM books WHERE COALESCE(last_accessed_at, created_at) < ?")) {
            ps.setLong(1, staleBefore);
            return ps.executeUpdate();
        }
    }

    public synchronized Stats stats() throws SQLException {
        int total = count("SELECT COUNT(*) FROM books");
        Long lastAccessedAt = maxNullable("SELECT MAX(last_accessed_at) FROM books");
        return new Stats(total, lastAccessedAt);
    }

    private BookRecord findByHash(String hash) throws SQLException {
        try (PreparedStatement ps = connection.prepareStatement(
                "SELECT * FROM books WHERE content_hash = ? ORDER BY created_at ASC LIMIT 1")) {
            ps.setString(1, hash);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next() ? read(rs) : null;
            }
        }
    }

    private int count(String sql) throws SQLException {
        try (Statement st = connection.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            return rs.next() ? rs.getInt(1) : 0;
        }
    }

    private Long maxNullable(String sql) throws SQLException {
        try (Statement st = connection.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            if (!rs.next()) return null;
            long value = rs.getLong(1);
            return rs.wasNull() ? null : value;
        }
    }

    private BookRecord read(ResultSet rs) throws SQLException {
        List<String> pages = GSON.fromJson(rs.getString("pages_json"),
                new TypeToken<List<String>>() {}.getType());
        long lastAccessed = rs.getLong("last_accessed_at");
        Long lastAccessedAt = rs.wasNull() ? null : lastAccessed;
        return new BookRecord(
                rs.getString("id"),
                rs.getString("content_hash"),
                rs.getString("title_raw"),
                rs.getString("author"),
                pages == null ? List.of("") : pages,
                rs.getLong("created_at"),
                rs.getLong("updated_at"),
                lastAccessedAt
        );
    }

    private void ensureColumn(Statement st, String column, String alterSql) throws SQLException {
        if (hasColumn(st, column)) return;
        st.execute(alterSql);
    }

    /**
     * Migrates v2 schemas (which had a soft-delete column) to the current model:
     * previously soft-deleted records are physically removed and the column dropped.
     * SQLite 3.35+ supports {@code ALTER TABLE ... DROP COLUMN}; sqlite-jdbc 3.45 ships it.
     */
    private void dropDeletedAtColumn(Statement st) throws SQLException {
        if (!hasColumn(st, "deleted_at")) return;
        st.execute("DELETE FROM books WHERE deleted_at IS NOT NULL");
        st.execute("DROP INDEX IF EXISTS idx_books_deleted");
        st.execute("ALTER TABLE books DROP COLUMN deleted_at");
    }

    private boolean hasColumn(Statement st, String column) throws SQLException {
        try (ResultSet rs = st.executeQuery("PRAGMA table_info(books)")) {
            while (rs.next()) {
                if (column.equalsIgnoreCase(rs.getString("name"))) return true;
            }
        }
        return false;
    }

    public record Stats(int total, Long lastAccessedAt) {}
}

File: BookLite/src/main/java/org/cubexmc/booklite/storage/BookRepository.kt
package org.cubexmc.booklite.storage

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.File
import java.sql.Connection
import java.sql.DriverManager
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Statement
import java.sql.Types
import java.util.Locale
import java.util.logging.Level
import org.cubexmc.booklite.BookLitePlugin
import org.cubexmc.booklite.config.ConfigManager
import org.cubexmc.booklite.model.BookRecord

class BookRepository(
    private val plugin: BookLitePlugin,
    private val config: ConfigManager,
) {
    private var connection: Connection? = null

    @Synchronized
    fun init() {
        try {
            Class.forName("org.sqlite.JDBC")
            if (!plugin.dataFolder.exists() && !plugin.dataFolder.mkdirs()) {
                plugin.logger.warning("Cannot create BookLite data folder.")
            }
            val db = File(plugin.dataFolder, config.getSqliteFile())
            connection = DriverManager.getConnection("jdbc:sqlite:${db.absolutePath}")
            requireConnection().createStatement().use { statement ->
                statement.execute("PRAGMA busy_timeout = 5000")
                if (config.isWal()) {
                    statement.execute("PRAGMA journal_mode = WAL")
                    statement.execute("PRAGMA synchronous = NORMAL")
                }
                statement.execute("PRAGMA foreign_keys = ON")
                statement.execute(
                    """
                    CREATE TABLE IF NOT EXISTS books (
                      id TEXT PRIMARY KEY,
                      content_hash TEXT NOT NULL,
                      title_raw TEXT NOT NULL,
                      author TEXT NOT NULL,
                      pages_json TEXT NOT NULL,
                      total_pages INTEGER NOT NULL,
                      created_at INTEGER NOT NULL,
                      updated_at INTEGER NOT NULL,
                      last_accessed_at INTEGER NULL
                    )
                    """.trimIndent(),
                )
                ensureColumn(statement, "last_accessed_at", "ALTER TABLE books ADD COLUMN last_accessed_at INTEGER NULL")
                dropDeletedAtColumn(statement)
                statement.execute("CREATE INDEX IF NOT EXISTS idx_books_hash ON books(content_hash)")
                statement.execute("CREATE INDEX IF NOT EXISTS idx_books_created ON books(created_at)")
                statement.execute("CREATE INDEX IF NOT EXISTS idx_books_last_accessed ON books(last_accessed_at)")
                statement.execute("PRAGMA user_version = 3")
            }
        } catch (exception: Exception) {
            plugin.logger.log(Level.SEVERE, "Failed to initialize BookLite SQLite storage.", exception)
            throw IllegalStateException(exception)
        }
    }

    @Synchronized
    fun close() {
        val currentConnection = connection ?: return
        try {
            currentConnection.close()
        } catch (exception: SQLException) {
            plugin.logger.log(Level.WARNING, "Failed to close BookLite database.", exception)
        }
    }

    @Synchronized
    @Throws(SQLException::class)
    fun saveOrGet(draft: BookRecord): BookRecord {
        val existing = findByHash(draft.contentHash())
        if (existing != null) {
            return existing
        }

        requireConnection().prepareStatement(
            """
            INSERT INTO books
              (id, content_hash, title_raw, author, pages_json, total_pages, created_at, updated_at, last_accessed_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """.trimIndent(),
        ).use { statement ->
            statement.setString(1, draft.id())
            statement.setString(2, draft.contentHash())
            statement.setString(3, draft.title())
            statement.setString(4, draft.author())
            statement.setString(5, GSON.toJson(draft.pages()))
            statement.setInt(6, draft.totalPages())
            statement.setLong(7, draft.createdAt())
            statement.setLong(8, draft.updatedAt())
            val lastAccessedAt = draft.lastAccessedAt()
            if (lastAccessedAt == null) {
                statement.setNull(9, Types.BIGINT)
            } else {
                statement.setLong(9, lastAccessedAt)
            }
            statement.executeUpdate()
        }
        return draft
    }

    @Synchronized
    @Throws(SQLException::class)
    fun find(id: String?): BookRecord? {
        if (id.isNullOrBlank()) {
            return null
        }
        requireConnection().prepareStatement("SELECT * FROM books WHERE id = ?").use { statement ->
            statement.setString(1, id)
            statement.executeQuery().use { resultSet ->
                return if (resultSet.next()) read(resultSet) else null
            }
        }
    }

    @Synchronized
    @Throws(SQLException::class)
    fun findByPrefix(prefix: String?, limit: Int): List<BookRecord> {
        val out = ArrayList<BookRecord>()
        if (prefix.isNullOrBlank()) {
            return out
        }
        requireConnection().prepareStatement(
            "SELECT * FROM books WHERE id LIKE ? ORDER BY created_at DESC LIMIT ?",
        ).use { statement ->
            statement.setString(1, "$prefix%")
            statement.setInt(2, maxOf(1, limit))
            statement.executeQuery().use { resultSet ->
                while (resultSet.next()) {
                    out.add(read(resultSet))
                }
            }
        }
        return out
    }

    @Synchronized
    @Throws(SQLException::class)
    fun completeIdsByPrefix(prefix: String?, limit: Int): List<String> {
        val out = ArrayList<String>()
        requireConnection().prepareStatement(
            "SELECT id FROM books WHERE id LIKE ? ORDER BY created_at DESC LIMIT ?",
        ).use { statement ->
            statement.setString(1, (prefix ?: "").lowercase(Locale.getDefault()) + "%")
            statement.setInt(2, maxOf(1, limit))
            statement.executeQuery().use { resultSet ->
                while (resultSet.next()) {
                    out.add(resultSet.getString("id"))
                }
            }
        }
        return out
    }

    @Synchronized
    @Throws(SQLException::class)
    fun list(offset: Int, limit: Int): List<BookRecord> {
        val out = ArrayList<BookRecord>()
        requireConnection().prepareStatement(
            "SELECT * FROM books ORDER BY created_at DESC LIMIT ? OFFSET ?",
        ).use { statement ->
            statement.setInt(1, maxOf(1, limit))
            statement.setInt(2, maxOf(0, offset))
            statement.executeQuery().use { resultSet ->
                while (resultSet.next()) {
                    out.add(read(resultSet))
                }
            }
        }
        return out
    }

    @Synchronized
    @Throws(SQLException::class)
    fun delete(id: String): Boolean {
        requireConnection().prepareStatement("DELETE FROM books WHERE id = ?").use { statement ->
            statement.setString(1, id)
            return statement.executeUpdate() > 0
        }
    }

    @Synchronized
    @Throws(SQLException::class)
    fun markAccessed(id: String, now: Long): Boolean {
        requireConnection().prepareStatement(
            "UPDATE books SET last_accessed_at = ?, updated_at = ? WHERE id = ?",
        ).use { statement ->
            statement.setLong(1, now)
            statement.setLong(2, now)
            statement.setString(3, id)
            return statement.executeUpdate() > 0
        }
    }

    /**
     * Permanently deletes records whose last player access (or, if never accessed,
     * creation time) is older than [staleBefore]. Admin operations like
     * /booklite read do not touch last_accessed_at, so they don't keep records alive.
     */
    @Synchronized
    @Throws(SQLException::class)
    fun purgeStale(staleBefore: Long): Int {
        requireConnection().prepareStatement(
            "DELETE FROM books WHERE COALESCE(last_accessed_at, created_at) < ?",
        ).use { statement ->
            statement.setLong(1, staleBefore)
            return statement.executeUpdate()
        }
    }

    @Synchronized
    @Throws(SQLException::class)
    fun stats(): Stats {
        val total = count("SELECT COUNT(*) FROM books")
        val lastAccessedAt = maxNullable("SELECT MAX(last_accessed_at) FROM books")
        return Stats(total, lastAccessedAt)
    }

    @Throws(SQLException::class)
    private fun findByHash(hash: String): BookRecord? {
        requireConnection().prepareStatement(
            "SELECT * FROM books WHERE content_hash = ? ORDER BY created_at ASC LIMIT 1",
        ).use { statement ->
            statement.setString(1, hash)
            statement.executeQuery().use { resultSet ->
                return if (resultSet.next()) read(resultSet) else null
            }
        }
    }

    @Throws(SQLException::class)
    private fun count(sql: String): Int {
        requireConnection().createStatement().use { statement ->
            statement.executeQuery(sql).use { resultSet ->
                return if (resultSet.next()) resultSet.getInt(1) else 0
            }
        }
    }

    @Throws(SQLException::class)
    private fun maxNullable(sql: String): Long? {
        requireConnection().createStatement().use { statement ->
            statement.executeQuery(sql).use { resultSet ->
                if (!resultSet.next()) {
                    return null
                }
                val value = resultSet.getLong(1)
                return if (resultSet.wasNull()) null else value
            }
        }
    }

    @Throws(SQLException::class)
    private fun read(resultSet: ResultSet): BookRecord {
        val pages: List<String>? = GSON.fromJson(
            resultSet.getString("pages_json"),
            object : TypeToken<List<String>>() {}.type,
        )
        val lastAccessed = resultSet.getLong("last_accessed_at")
        val lastAccessedAt = if (resultSet.wasNull()) null else lastAccessed
        return BookRecord(
            resultSet.getString("id"),
            resultSet.getString("content_hash"),
            resultSet.getString("title_raw"),
            resultSet.getString("author"),
            pages ?: listOf(""),
            resultSet.getLong("created_at"),
            resultSet.getLong("updated_at"),
            lastAccessedAt,
        )
    }

    @Throws(SQLException::class)
    private fun ensureColumn(statement: Statement, column: String, alterSql: String) {
        if (hasColumn(statement, column)) {
            return
        }
        statement.execute(alterSql)
    }

    /**
     * Migrates v2 schemas (which had a soft-delete column) to the current model:
     * previously soft-deleted records are physically removed and the column dropped.
     * SQLite 3.35+ supports `ALTER TABLE ... DROP COLUMN`; sqlite-jdbc 3.45 ships it.
     */
    @Throws(SQLException::class)
    private fun dropDeletedAtColumn(statement: Statement) {
        if (!hasColumn(statement, "deleted_at")) {
            return
        }
        statement.execute("DELETE FROM books WHERE deleted_at IS NOT NULL")
        statement.execute("DROP INDEX IF EXISTS idx_books_deleted")
        statement.execute("ALTER TABLE books DROP COLUMN deleted_at")
    }

    @Throws(SQLException::class)
    private fun hasColumn(statement: Statement, column: String): Boolean {
        statement.executeQuery("PRAGMA table_info(books)").use { resultSet ->
            while (resultSet.next()) {
                if (column.equals(resultSet.getString("name"), ignoreCase = true)) {
                    return true
                }
            }
        }
        return false
    }

    private fun requireConnection(): Connection =
        requireNotNull(connection) { "BookLite database connection is not initialized" }

    class Stats(
        private val total: Int,
        private val lastAccessedAt: Long?,
    ) {
        fun total(): Int = total

        fun lastAccessedAt(): Long? = lastAccessedAt
    }

    private companion object {
        val GSON: Gson = Gson()
    }
}
